### PR TITLE
Integrate vm improve 1.5.5

### PR DESCRIPTION
**Pull request by sasurobert**
## Reasoning behind the pull request
Enable more opcodes and options for developers.
  
## Proposed changes
Integrate new crypto endpoints - BLS MultiSig, Secp256R1
Integrate new Opcodes for Relayer, OriginalCaller
Integrate TransferAndExecuteByUser for SovereignChains
Integrate EGLD in MultiTransfer

## Testing procedure
System test every aspect noted before.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: process/errors.go [hunk 1]
```diff
@@ -1226,3 +1226,6 @@ var ErrInvalidAsyncArguments = errors.New("invalid arguments to process async/ca
 
 // ErrNilSentSignatureTracker defines the error for setting a nil SentSignatureTracker
 var ErrNilSentSignatureTracker = errors.New("nil sent signature tracker")
+
+// ErrTransferAndExecuteByUserAddressesIsNil signals that transfer and execute by user addresses are nil
+var ErrTransferAndExecuteByUserAddressesIsNil = errors.New("transfer and execute by user addresses are nil")
```
Comment (mariusmihaic) [inline process/errors.go:1231]:
```suggestion
// ErrTransferAndExecuteByUserAddressesAreNil signals that transfer and execute by user addresses are nil
var ErrTransferAndExecuteByUserAddressesAreNil = errors.New("transfer and execute by user addresses are nil")
```

File: process/factory/shard/vmContainerFactory_test.go [hunk 2]
```diff
@@ -137,6 +139,33 @@ func TestNewVMContainerFactory_NilHasherShouldErr(t *testing.T) {
 	assert.Equal(t, process.ErrNilHasher, err)
 }
 
+func TestNewVMContainerFactory_NilPubKeyConverterShouldErr(t *testing.T) {
```
Comment (mariusmihaic) [inline process/factory/shard/vmContainerFactory_test.go:142]: Perhaps we could add a working test to test that the decoded addresses are correctly saved internally in that map?

File: common/constants.go [hunk 3]
```diff
@@ -1015,5 +1015,7 @@ const (
 	StakingV4StartedFlag                               core.EnableEpochFlag = "StakingV4StartedFlag"
 	AlwaysMergeContextsInEEIFlag                       core.EnableEpochFlag = "AlwaysMergeContextsInEEIFlag"
 	DynamicESDTFlag                                    core.EnableEpochFlag = "DynamicEsdtFlag"
+	EGLDInESDTMultiTransferFlag                        core.EnableEpochFlag = "EGLDInESDTMultiTransferFlag"
+	CryptoAPIV2Flag                                    core.EnableEpochFlag = "CryptoAPIV2Flag"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply common/constants.go:1019]: this is not API/ but OPCODEs. Now I am thinking about renaming these.

File: common/enablers/enableEpochsHandler_test.go [hunk 4]
```diff
@@ -440,6 +442,8 @@ func TestEnableEpochsHandler_GetActivationEpoch(t *testing.T) {
 	require.Equal(t, cfg.StakingV4Step1EnableEpoch, handler.GetActivationEpoch(common.StakingV4StartedFlag))
 	require.Equal(t, cfg.AlwaysMergeContextsInEEIEnableEpoch, handler.GetActivationEpoch(common.AlwaysMergeContextsInEEIFlag))
 	require.Equal(t, cfg.DynamicESDTEnableEpoch, handler.GetActivationEpoch(common.DynamicESDTFlag))
+	require.Equal(t, cfg.EGLDInMultiTransferEnableEpoch, handler.GetActivationEpoch(common.EGLDInESDTMultiTransferFlag))
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply common/enablers/enableEpochsHandler_test.go:445]: added the same tests as we had for other flags.

File: config/tomlConfig_test.go [hunk 5]
```diff
@@ -199,6 +203,11 @@ func TestTomlParser(t *testing.T) {
             { StartEpoch = 12, Version = "v0.3" },
             { StartEpoch = 88, Version = "v1.2" },
         ]
+		TransferAndExecuteByUserAddresses = [ # TODO: set real contract addresses
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply config/tomlConfig_test.go:206]: I just copy pasted the TODO.

File: factory/api/apiResolverFactory_test.go [hunk 6]
```diff
@@ -186,7 +186,8 @@ func TestCreateApiResolver(t *testing.T) {
 		failingStepsInstance.addressPublicKeyConverterFailingStep = 3
 		apiResolver, err := api.CreateApiResolver(failingArgs)
 		require.NotNil(t, err)
-		require.True(t, strings.Contains(strings.ToLower(err.Error()), "public key converter"))
+		fmt.Println(err.Error())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply factory/api/apiResolverFactory_test.go:189]: done

File: process/errors.go (hunk 1 above)
Comment (sasurobert) [reply process/errors.go:1231]: done

File: process/factory/shard/vmContainerFactory_test.go (hunk 2 above)
Comment (sasurobert) [reply process/factory/shard/vmContainerFactory_test.go:142]: ok. doing it.

File: integrationTests/vm/txsFee/apiTransactionEvaluator_test.go [hunk 7]
```diff
@@ -30,26 +31,25 @@ func TestSCCallCostTransactionCost(t *testing.T) {
...
+	gasLimit := uint64(10000000)
 
 	_, _ = vm.CreateAccount(testContext.Accounts, sndAddr, 0, senderBalance)
 
 	tx := vm.CreateTransaction(0, big.NewInt(0), sndAddr, scAddress, gasPrice, gasLimit, []byte("increment"))
 
 	res, err := testContext.TxCostHandler.ComputeTransactionGasLimit(tx)
 	require.Nil(t, err)
-	require.Equal(t, uint64(418), res.GasUnits)
+	fmt.Println(res.GasUnits)
```
Comment (BeniaminDrasovean) [inline integrationTests/vm/txsFee/apiTransactionEvaluator_test.go:51]: Remove this print.

File: testscommon/vmcommonMocks/userAccountStub.go [hunk 8]
```diff
@@ -74,6 +75,14 @@ func (uas *UserAccountStub) AddToBalance(value *big.Int) error {
 	return nil
 }
 
+// SubFromBalance -
+func (uas *UserAccountStub) SubFromBalance(value *big.Int) error {
+	if uas.AddToBalanceCalled != nil {
```
Comment (BeniaminDrasovean) [inline testscommon/vmcommonMocks/userAccountStub.go:80]: SubFromBalanceCalled

Comment (sasurobert) [reply testscommon/vmcommonMocks/userAccountStub.go:80]: done

File: integrationTests/vm/txsFee/apiTransactionEvaluator_test.go (hunk 7 above)
Comment (sasurobert) [reply integrationTests/vm/txsFee/apiTransactionEvaluator_test.go]: done

**Review by BeniaminDrasovean — DISMISSED** [review]

**Review by mariusmihaic — APPROVED** [review]

**Review by gabi-vuls — DISMISSED** [review]
Normal allin test: v1.6.15-dev-config-eb2e06c06d -> integrate-vm-improve-1.5.5-e26260c87e

--- Specific errors ---

block hash does not match 401
wrong nonce in block 291
miniblocks does not match 0
num miniblocks does not match 0
miniblock hash does not match 0
block bodies does not match 0
receipts hash missmatch 0

/------/


--- Statistics ---

Nr. of all ERRORS:  0
Nr. of all WARNS:  230
Nr. of new ERRORS:  0
Nr. of new WARNS:  0
Nr. of PANICS:  0

/------/

--- ERRORS ---

/------/

--- WARNINGS ---

/------/

**Review by BeniaminDrasovean — APPROVED** [review]

**Pull request merged**